Save and load CallMap from disk

lambdageek opened the pull request:
Fixes #11

lambdageek committed:
Add DumpCallMap module to write callmap to a file

lambdageek committed:
new OutputAction datatype; refactor main

* The OutputAction is a command-line argument that determines what we should
  do (right now the only thing we can do is run the analysis) - the output mode
  specifies the format for the analysis results.

* Refactor main to separate argument, config and C source parsing from running
  the analysis.

lambdageek committed:
Add new --mode=graph arg; new GraphAction; dumpCallGraph function

With GraphAction we don't run any analysis, just parse the C sources build the
call graph and emit it on stdout

lambdageek committed:
Add ParseCallMap module to read CallMap from a file

lambdageek committed:
Add ability to analyze .graph files and .c files together

lambdageek commented:
It occured to me just now that we don't really care about the file format - I wonder if I should've experimented with aeson and saved myself a whole lot of typing.  (Although some of the `language-c` data types probably would've needed handwritten `ToJSON/FromJSON` instances so it wouldn't have been completely trivial in either case)

lambdageek commented:
There's not great UX around this right now.

1. Generate a graph file by running `ward CPP --config=foo.config --mode=graph inputs.[c|graph] > result.graph`
2. Consume graphs by running `ward CPP --config=foo.config --mode=[compiler|html] inputs.[c|graph] ...`

* Any file with a `.graph` extension is assumed to be a callmap input, otherwise it's a C source file input.

* With `--mode=graph` we write progress messages to `stderr` and the resulting graph to `stdout`.  In analysis mode, if we're emitting HTML we don't write progress messages and write the analysis results to `stdout`, otherwise we write progress messages and analysis results to `stdout`.

* The only reason we need a config when generating a callgraph is to validate the permissions.  We should probably just record them as given in the graph and only validate when running the analysis.  Or maybe not.  I'm not sure what works nicer.

lambdageek committed:
[Check.Permissions] Break up monolithic process function a bit

lambdageek commented:
About e1ad464 -  I broke up the `process` function into some smaller pieces - there's really not that much global environment to carry around and I think it's a bit clearer what's going on this way.  Feel free to drop it if you don't like it.

lambdageek committed:
[Check.Permissions] Data.Tree folds in flattened order

The call to Tree.flatten is not needed - the loop iterates in the same order.

lambdageek committed:
[Check.Permissions] Pull per-node loop body out of inferPermissionsSCC

lambdageek committed:
[Check.Permissions] Pull out [PermissionAction] -> Site pure function

lambdageek committed:
[Check.Permissions] pull permissionsFromCallSites out of propagatePermissionsNode

lambdageek committed:
[Check.Permissions] Pull derivePermissionActions out of permissionsFromCallSites

lambdageek committed:
[Check.Permissions] Fix incorrect termination condition (Fixes #15)

Previously, we overwrote 'growing :: IORef Bool' with the result of processing
the last graph node.  That means if a prior node changed, but the last one did
not, we wouldn't go around for another iteration.

evincarofautumn commented on src/DumpCallMap.hs:
```diff
@@ -0,0 +1,126 @@
+
+encodeCallMapItem :: Ident -> CallMapItem -> B.Builder
+encodeCallMapItem f (ni, tree, ps) =
+  (form "function" False $ encodeIdent f <> space <> encodeNodeInfo ni <> newline
```
This would be nicer with `mconcat`.

evincarofautumn commented on src/DumpCallMap.hs:
```diff
@@ -0,0 +1,126 @@
+
+encodePositionSpan :: Pos.Position -> Pos.PosLength -> B.Builder
+encodePositionSpan pstart pl =
+  case () of
```
This could be replaced with just guards.

evincarofautumn commented on src/DumpCallMap.hs:
```diff
@@ -0,0 +1,126 @@
+encodePerms = form "actions" False . mconcat . intersperse space . map encodePermissionAction . toList
+
+encodePermissionAction :: PermissionAction -> B.Builder
+encodePermissionAction pa =
```
```
encodePermissionAction = go . permissionActionName
  where go = form "need" False . encodePermissionName
```

evincarofautumn commented on src/ParseCallMap.hs:
```diff
@@ -0,0 +1,313 @@
+{-# language FlexibleInstances, OverloadedStrings, ViewPatterns, GeneralizedNewtypeDeriving #-}
+module ParseCallMap (fromFile, fromSource, CallMapParseError) where
+
+import Prelude hiding (exp)
```
This `hiding` seems unnecessary. For the prelude it’s usually fine, but I try to avoid `hiding` because it breaks when dependencies remove exports.

lambdageek replied:
ghc complained - although it looks like that was a previous iteration of this code that defined and `exp` expression parser which eventually became `theForm`

evincarofautumn commented on src/ParseCallMap.hs:
```diff
@@ -0,0 +1,313 @@
+
+functionForm :: Parser (Ident, CallMapItem)
+functionForm = theForm "function" ((,)
+                                    <$> identForm
```
I try to avoid alignment like this because it’s not great for editing & diffing.

lambdageek replied:
Just so we're on the same page - the fact that the alignment of the lines below depends on the length of everything to the left of `(,)` is the issue, yea?  ok.

evincarofautumn replied:
Yup

evincarofautumn commented on src/ParseCallMap.hs:
```diff
@@ -0,0 +1,313 @@
+seqForest :: Parser (CallTree Ident)
+seqForest = joinForest <$> many subcallForm
+  where
+    joinForest :: [CallTree a] -> CallTree a
```
This is general enough to go in `Types`.

lambdageek replied:
maybe.  it's really the `mconcat` of the `Sequence`/`Nop` monoid structure for `CallTree`. However I didn't feel great about writing the `Monoid` instance because I couldn't convince myself that the `Choice`/`Nop` monoid is clearly less useful.  In any case I was going to spend tomorrow playing around with a sequence-of-choices (ie two data types) presentation of call trees to see if I liked it better in `Check.Permissions` in which case the monoid structure would be more natural and this function would go away.

evincarofautumn replied:
Makes sense. Modelling algebraic structures with typeclasses always felt a little awkward to me. It forces us to phrase instances as “T is a monoid”, which doesn’t make much sense. What I’d really like is something like [this C++ representation](https://gist.github.com/evincarofautumn/2b5f004ca81e33c62ff0):

```
class Monoid (m :: *) (mappend :: m -> m -> m) (mempty :: m)
instance Monoid Int '(+) '0
instance Monoid Int '(*) '1

class Monoid (m :: *) (mappend :: m -> m -> m) where mempty :: m
instance Monoid Int '(+) where mempty = 0
instance Monoid Int '(*) where mempty = 1
```

evincarofautumn commented on src/ParseCallMap.hs:
```diff
@@ -0,0 +1,313 @@
+    kw :: (Sidentifier, (PermissionName -> PermissionAction)) -> Parser (PermissionName -> PermissionAction)
+    kw (ident,f) = f <$ theIdentifier ident
+    actions :: [(Sidentifier, PermissionName -> PermissionAction)]
+    actions = [ ("need", Need)
```
Excessive alignment.

lambdageek replied:
Not sure what "excessive alignment" means - you wrote that in a couple of places - could you pick one and show me your preferred way of writing it?

evincarofautumn replied:
All of these refer to: https://github.com/evincarofautumn/Ward/pull/13#discussion_r126854838

Generally I avoid this:

```
xxxxx xx xx = do xxx xx
                 xx
                 xxxx xxxxx xxxx
                 xxxxxx x

x xx x = xxx xxxxx xxxxx
       + xxxxxxxxxx xxxx

xxx = [ xxx
      , xxxxx
      , xx xxxx
      , x
      ]

xxxxxxx xxx x xxxxx = xx (xxxxxxx xxxxx
                        * xxxxxxx xxxxx
                        + xxxx xx x)
```

And prefer this:

```
-- Newline after do, let, (case) of, &c.
xxxxx xx xx = do
  xxx xx
  xx
  xxxx xxxxx xxxx
  xxxxxx x

-- Operator at beginning of line…
x xx x
  = xxx xxxxx xxxxx
  + xxxxxxxxxx xxxx

-- …unless it saves you a level of indentation.
xxx =
  [ xxx
  , xxxxx
  , xx xxxx
  , x
  ]

-- Just add an indent level, don’t align with parentheses.
xxxxxxx xxx x xxxxx = xx (xxxxxxx xxxxx
  * xxxxxxx xxxxx
  + xxxx xx x)
```

lambdageek replied:
Okay, most of these are what I try to do.

But this last one - oh god, seriously?
```
-- Just add an indent level, don’t align with parentheses.
xxxxxxx xxx x xxxxx = xx (xxxxxxx xxxxx
  * xxxxxxx xxxxx
  + xxxx xx x)
```

I think I'll just write a `let` or `where` :smile:

evincarofautumn replied:
Well, thinking on it, I actually tend to do something more like these:

```
xxxxxxx xxx x xxxxx
  = xx (xxxxxxx xxxxx
    * xxxxxxx xxxxx
    + xxxx xx x)

xxxxxxx xxx x xxxxx
  = xx
    (xxxxxxx xxxxx
      * xxxxxxx xxxxx
      + xxxx xx x)
```

Or give it a name, yeah.

evincarofautumn commented on src/ParseCallMap.hs:
```diff
@@ -0,0 +1,313 @@
+  (NoPosCSpan <$ theIdentifier "noPos")
+  <|> (BuiltinPosCSpan <$ theIdentifier "builtin")
+  <|> (InternalPosCSpan <$ theIdentifier "internal")
+  <|> theForm "span" (mkSpanPos
```
Excessive alignment.

evincarofautumn commented on src/ParseCallMap.hs:
```diff
@@ -0,0 +1,313 @@
+
+sourceForm :: Parser CPos.Position
+sourceForm =
+  theForm "source" (mkPos
```
Excessive alignment.

evincarofautumn commented on src/ParseCallMap.hs:
```diff
@@ -0,0 +1,313 @@
+extractSpan = snd
+
+-- unused
+-- | @Spanning a@ is a comonad with 'extendSpan' its composition operation and
```
Nice observation. I might actually make use of that in another project.

lambdageek replied:
The original history of this module was that it defined an intermediate representation of completely uninterpreted S-expressions and then traversed those trees to construct the `CallMap`.  So the S-expression nodes had `Spanning Sexp` children so I could give error messages with line numbers.

In the end I scrapped that whole approach because `Parsec`s 4-continuation error monad gives much better error messages than what I could cook up with `ExceptT` and because the code to just parse the forms I need directly is like half the size.

Which is all a roundabout way of saying that `Spanning` worked out great and I recommend something like it, but it's pretty extraneous here.

evincarofautumn commented on src/ParseCallMap.hs:
```diff
@@ -0,0 +1,313 @@
+  return (f n)
+
+sign :: Parser (Integer -> Integer)
+sign = (char '-' >> return negate)
```
```
sign = choice [negate <$ char '-', id <$ char '+', pure id]
```

evincarofautumn commented on src/ParseCallMap.hs:
```diff
@@ -0,0 +1,313 @@
+  <?> ""
+
+decimal     = number 10 digit
+hexadecimal = oneOf "xX" *> number 16 hexDigit
```
Why are `hexadecimal` and `octal` necessary?

evincarofautumn commented on src/Graph.hs:
```diff
@@ -33,7 +33,24 @@ pattern WardKeyword <- Ident WardKeywordStr _hash _ni
+fromProcessingUnits config units =
+  let (cs, gs) = partitionProcessingUnits units
+      g0 :: CallMap
+      g0 = mconcat (map snd gs)
```
`mconcat . map f` = `foldMap f`

evincarofautumn commented on src/Main.hs:
```diff
@@ -8,12 +8,13 @@ import Config
 import Language.C.Syntax.AST (CTranslUnit)
 import Language.C.System.GCC (newGCC)
 import System.Exit (exitFailure)
+import qualified System.FilePath as FP
```
I try not to abbreviate import aliases, usually taking the last name part (`FilePath`). But it seems like you just need `takeExtension` here.

evincarofautumn commented on src/Main.hs:
```diff
@@ -40,12 +42,7 @@ main = do
-    in forM (Args.translationUnitPaths args)
-      $ parseCFile preprocessor temporaryDirectory
-      $ Args.preprocessorFlags args
+  parseResults <- traverse (parseInput args) (Args.translationUnitPaths args)
```
I had to stop myself from suggesting `liftA2 traverse parseInput Args.translationUnitPaths args`. I may have a point-free problem.

evincarofautumn commented on src/Main.hs:
```diff
@@ -106,9 +104,35 @@ logProgress args s = case Args.outputAction args of
+parseInput args path =
+  case classifyPath path of
+    CSourcePathClass ->
+      (either (Left . CSourceUnitParseError) (Right . CSourceProcessingUnit)) <$> parseCInput args path
```
`either (Left . f) (Right . g)` = `bimap f g`. This pattern shows up in a few places.

lambdageek replied:
Ooh nice. I always forget about `bifunctors`

evincarofautumn commented on src/Check/Permissions.hs:
```diff
@@ -349,31 +349,15 @@ inferPermissionsSCC implicitPermissions graphLookup graphVertex scc = do
-            Waive{} -> mempty
+        -- We initialize the first call site of the function according to its
+        -- permission actions.
+        IOVector.modify sites (\old -> old <> newInitialSite) 0
```
`(<> newInitialSite)`

lambdageek replied:
I really dislike operator sections.

evincarofautumn reviewed (commented):
This is looking good. I’d suggest adding a couple of smoke tests:

* Running analysis on a graph generated from a source file should produce the same results as the source file. (I guess we could do this for all tests?)
* `{Dump,Parse}CallMap` should roundtrip

Now that you mention it, aeson would probably have worked just fine—the only reason I didn’t use JSON for the config file format is that I wanted a structured format that was convenient to edit by hand, which isn’t really relevant for the graph dumps. But it’s probably not worth changing at this point.

The workflow you describe seems fine to me, or at least it’s what I had in mind originally. We can improve things as we go.

There are some stylistic differences that I don’t care very strongly about, but for future reference, I generally follow [Johan Tibell’s Haskell Style Guide](https://github.com/tibbe/haskell-style-guide/blob/master/haskell-style.md) except:

* 2-space indents
* Avoid anything that requires fiddly formatting work (alignment, sorting imports other than lexicographically)
* Space after lambda
* Break lines before operators unless you can avoid a level of indentation by breaking after
* Prefer point-free style unless it looks like trash

I left a few comments about minor issues. Feel free to merge (no merge commit please) when you feel like this is ready.

lambdageek committed:
[tests] Add callmap roundtrip tests for all the .c files in test/

Check that serializing the call map from a parsed C file and reading it back in
is == to the original callmap.
